### PR TITLE
Feature/szewczyk/refactor external events

### PR DESCRIPTION
**Pull request by robszewczyk**
Rewrite the storage of external events

Previously, the external events were stored in dedicated slots attached to each event buffer.  This created a number of problems:

* the number of slots was fixed and set across the different importance levels; as a result, we made rather inefficient use of that memory.
* the events could have been unregistered, but their slot was not freed until that range of event IDs was rotated out of the buffer. this behavior was rather counterintuitive, and did not leave the app developers with the control they wanted.

This patch replaces the dedicated slots for external events with sentinel entries for external events in the regular event buffers. The result is that the effective number of slots is bounded by the
event buffer sizes rather than additional configuration parameters, and external events are flushed the same way as all other event entries.  The existing API for external events is modified slightly: instead of passing around pointers as handles to the external events, we identify external event blocks by the last event ID in the sequence.  The full structure of `ExternalEvents` remains accessible within the `FetchExternalEventsFunct` as an additional member of the `EventLoadOutContext`.

External event structure needs to be modified when the external events are unregistered.  This created a need create an API that would position a reader at the right place in the TLV buffer.  Additionally, we needed a `WeaveCircularTLVBuffer` constructor that would start the circular buffer at any point in the buffer (as opposed to the 0-offset).  Those changes are presented here as individual separable commits; I would be happy to present them as smaller more digestible pull requests.

### CODE REVIEW

File: src/test-apps/MockExternalEvents.cpp [hunk 1]
```diff
@@ -110,10 +110,11 @@ WEAVE_ERROR FetchMockExternalEvents(EventLoadOutContext *aContext)
     int i;
     for (i = 0; i < 3; i++)
```
[the remark this answers is not in the document]
Comment (robszewczyk) [reply src/test-apps/MockExternalEvents.cpp]: Number of importance levels.

[the remark this answers is not in the document]
Comment (gerickson) [reply src/test-apps/MockExternalEvents.cpp]: Thanks. Perhaps not critical for this review, but the usual kImportanceType_Last - kImportanceType_First + 1 math would be good here.

[the remark this answers is not in the document]
Comment (gerickson) [reply src/test-apps/MockExternalEvents.cpp]: I'd recommend breaking this into two pull requests:

- Weave TLV changes + missing unit tests.
- Event logging changes that use those + unit tests.

[the remark this answers is not in the document]
Comment (robszewczyk) [reply src/test-apps/MockExternalEvents.cpp]: Will do.

[the remark this answers is not in the document]
Comment (robszewczyk) [reply src/test-apps/MockExternalEvents.cpp]: Rebased the PR against the merged PR#91, and made the external event support a configurable option, enabled in the standalone builds.

**Pull request merged**